internal/bufpushaction/bufpushaction.go

```diff
@@ -0,0 +1,260 @@
+)
+
+// Main is the entrypoint to the buf CLI.
+func Main(name string) {
+	appcmd.Main(context.Background(), NewRootCommand(name))
+}
+
+func NewRootCommand(name string) *appcmd.Command {
+	builder := appflag.NewBuilder(name, appflag.BuilderWithTimeout(120*time.Second))
+	return &appcmd.Command{
+		Use:   name + " <input> <track> <git-commit-hash>",
+		Short: "helper for the GitHub Action buf-push-action",
+		Args:  cobra.ExactArgs(3),
+		Run:   builder.NewRunFunc(run, interceptErrorForGithubAction),
+	}
+}
+
+func run(ctx context.Context, container appflag.Container) error {
+	input := container.Arg(0)
+	track := container.Arg(1)
+	currentGitCommit := container.Arg(2)
+
+	// exit early if buf or BUF_TOKEN isn't found
+	if _, err := exec.LookPath("buf"); err != nil {
+		return errors.New(`buf is not installed; please add the "bufbuild/buf-setup-action" step to your job found at https://github.com/bufbuild/buf-setup-action`)
+	}
+	if container.Env(bufTokenKey) == "" {
+		return errors.New("a buf authentication token was not provided")
+	}
+	githubRepository := container.Env(githubRepositoryKey)
+	if githubRepository == "" {
+		return errors.New("a github repository was not provided")
+	}
+	repoParts := strings.Split(githubRepository, "/")
+	if len(repoParts) != 2 {
+		return errors.New("a github repository was not provided in the format owner/repo")
+	}
+	githubClient, err := github.NewClient(ctx, container.Env(githubTokenKey), "buf-push-action", "", githubRepository)
+	if err != nil {
+		return err
+	}
+	bucket, err := storageos.NewProvider().NewReadWriteBucket(input)
+	if err != nil {
+		return fmt.Errorf("config file not found: %s", input)
+	}
+	moduleName, err := getNameFromConfigFile(ctx, bucket)
+	if err != nil {
+		return fmt.Errorf("name not found in  %s", input)
+	}
+	return pusher{
+		input:            input,
+		track:            track,
+		stdout:           container.Stdout(),
+		stderr:           container.Stderr(),
+		currentGitCommit: currentGitCommit,
+		moduleName:       moduleName,
+		githubClient:     githubClient,
+		bufRunner: &bufRunner{
+			bufToken: container.Env(bufTokenKey),
+			path:     container.Env("PATH"),
+		},
+	}.push(ctx)
+}
+
+// interceptErrorForGithubAction intercepts errors and wraps them in formatting required for an error to be shown in
+// the workflow results.
+func interceptErrorForGithubAction(
+	next func(context.Context, appflag.Container) error,
+) func(context.Context, appflag.Container) error {
+	return func(ctx context.Context, container appflag.Container) error {
+		err := next(ctx, container)
+		if err == nil {
+			return nil
+		}
+		return fmt.Errorf("::error::%v", err)
+	}
+}
+
+func getNameFromConfigFile(ctx context.Context, bucket storage.ReadBucket) (_ string, retErr error) {
+	file, err := bucket.Get(ctx, "buf.yaml")
+	if err != nil {
+		return "", err
+	}
+	defer func() {
+		retErr = multierr.Append(retErr, file.Close())
+	}()
+	name := struct {
+		Name string `yaml:"name"`
+	}{}
```

Comment:
rather than re-defining a (subset) of the buf yaml here, can we import the file format from `github.com/bufbuild/buf` somewhere?

Reply:
I looked into this, but it became much more complex than just parsing the yaml.